linkerd/distribute/src/service.rs

```diff
@@ -0,0 +1,317 @@
+        match assert_ready!(handle0.poll_request()) {
+            Some(((), rsp)) => rsp.send_response(()),
+            _ => panic!("expected request"),
+        }
+        assert_ready_ok!(call.poll());
+    }
+
+    #[test]
+    fn first_available_uses_second() {
+        let (mock0, mut handle0) = mock::pair();
+        let (mock1, mut handle1) = mock::pair();
+        let mut dist_svc = mock::Spawn::new(Distribute::new(
+            vec![("foo", mock0), ("bar", mock1)].into_iter().collect(),
+            Distribution::FirstAvailable(Arc::new(["foo", "bar"])),
+        ));
+        assert_eq!(dist_svc.get_ref().backends.len(), 2);
+
+        handle0.allow(0);
+        handle1.allow(1);
+        assert_ready_ok!(dist_svc.poll_ready());
+        assert_eq!(dist_svc.get_ref().ready_idx, Some(1));
+        let mut call = task::spawn(dist_svc.call(()));
+        match assert_ready!(handle1.poll_request()) {
+            Some(((), rsp)) => rsp.send_response(()),
+            _ => panic!("expected request"),
+        }
+        assert_ready_ok!(call.poll());
+    }
+
+    #[test]
+    fn random_available_woken() {
+        let (mock0, mut handle0) = mock::pair::<(), ()>();
+        let (mock1, mut handle1) = mock::pair::<(), ()>();
+        let mut dist_svc = mock::Spawn::new(Distribute::new(
+            vec![("foo", mock0), ("bar", mock1)].into_iter().collect(),
+            Distribution::random_available([("foo", 1), ("bar", 99999)]).unwrap(),
+        ));
+        assert_eq!(dist_svc.get_ref().backends.len(), 2);
+
+        handle0.allow(0);
+        handle1.allow(0);
+        assert_pending!(dist_svc.poll_ready());
+        handle0.allow(1);
+        assert!(dist_svc.is_woken());
+    }
+
+    #[test]
+    fn random_available_follows_weight() {
+        let (mock0, mut handle0) = mock::pair();
+        let (mock1, mut handle1) = mock::pair();
+        let mut dist_svc = mock::Spawn::new(Distribute::new(
+            vec![("foo", mock0), ("bar", mock1)].into_iter().collect(),
+            Distribution::random_available([("foo", 1), ("bar", 99999)]).unwrap(),
+        ));
+        assert_eq!(dist_svc.get_ref().backends.len(), 2);
+
+        handle0.allow(1);
+        handle1.allow(1);
+        assert_ready_ok!(dist_svc.poll_ready());
+        assert_eq!(dist_svc.get_ref().ready_idx, Some(1));
+        let mut call = task::spawn(dist_svc.call(()));
+        match assert_ready!(handle1.poll_request()) {
+            Some(((), rsp)) => rsp.send_response(()),
+            _ => panic!("expected request"),
+        }
+        assert_ready_ok!(call.poll());
+    }
+
+    #[test]
+    fn random_available_follows_availability() {
+        let (mock0, mut handle0) = mock::pair();
+        let (mock1, mut handle1) = mock::pair();
+        let mut dist_svc = mock::Spawn::new(Distribute::new(
+            vec![("foo", mock0), ("bar", mock1)].into_iter().collect(),
+            Distribution::random_available([("foo", 1), ("bar", 99999)]).unwrap(),
```

Comment:
i noticed all the `random_available` tests have two services...should we be testing cases with >2 services?

Reply:
Maybe? what cases do you think we're not covering in the above logic? It's already a little sketchy writing tests overweighted random distributions (this test will fail 0.00001% of the time).